### PR TITLE
[move-prover] Monomorphization analysis

### PR DESCRIPTION
**Pull request by wrwg**
This implements a new bytecode processor which performs monomorphization analysis. This analysis computes all type instantiations in a program for structs and functions which are inlined for verification.

The basic assumptions for monomorphization are the followings:

- We can verify top-level functions with generic parameters by fixing those to a "given" type. Those given types will later be introduced in the Boogie backend.
- For functions which are not opaque and therefore are inlined for verification, we need to transitively compute all instantiations (with concrete or formal given types from top-level functions).  For opaque functions, we expect that pre/post will be specialized by the backend at caller points, so we do not to compute their instantiations.
- All structs types used directly or indirectly need to have instantiations computed.
- For specification expressions and functions, we expect the backend to be able to represent them generically. (Boogie supports this and we do it already today.)

This PR also makes some some changes to the bytecode pipeline infra:

- As I first thought to use compositional analysis for this, I removed some restrictions there to work with function variants.
- However, I later decided to do this with a new style of processor, namely a "single run" processor. This type of processor does not visit function by function, but just performs one global analysis step.
- There is also now a facility for a processor to dump global (not per function) output.

The PR also fixes an issue with the new way how we collect sources and dependencies in move-lang. Specifically, we canonicalize file names such that `./foo.move` and `/<path/foo.move` are recognized as the same.

## Motivation

Monomorphization

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

New test

## Related PRs

NA

### CODE REVIEW

**Comment by wrwg** [comment]
@sblackshear mostly FYI as I updated compositional analysis but then did not use this update and instead introduced a new simpler style of global custom analysis; this may also be  more adequate for things like packed typed and writeset analysis.

File: language/move-lang/src/lib.rs
Comment (mengxu-fb) [inline language/move-lang/src/lib.rs:733]: Thanks a lot for the fix! Double checking with @tnowacki as the canonicalization is not done in the original intersection check, so just want to make sure that the "no canonicalization" is not by design.

File: language/move-prover/bytecode/src/function_target_pipeline.rs
Comment (mengxu-fb) [inline language/move-prover/bytecode/src/function_target_pipeline.rs:23]: I guess we could keep the `Copy` here but I am fine either way

File: language/move-prover/bytecode/tests/mono_analysis/test.exp
Comment (shazqadeer) [inline language/move-prover/bytecode/tests/mono_analysis/test.exp:72]: What does this <#0> mean?

File: language/move-prover/bytecode/src/mono_analysis.rs
Comment (mengxu-fb) [inline language/move-prover/bytecode/src/mono_analysis.rs:201]: I guess the convention is `add_type_` (a suffix underscore) to denote internal function

**Review by mengxu-fb — APPROVED** [review]
Accepting to unblock, especially for the fix on canonicalization.

File: language/move-prover/bytecode/tests/mono_analysis/test.exp
Comment (mengxu-fb) [reply language/move-prover/bytecode/tests/mono_analysis/test.exp:72]: My understanding is index of the type parameter

**Review by shazqadeer — APPROVED** [review]
I could not fully understand the golden output in test.exp (see my question there).  But I read through the code in mono_analysis.rs and it made sense to me.

I thought for a bit about the verification scheme you have in mind where a generic function is verified for an arbitrary uninterpreted type.  It seems correct to me intuitively although I could not come up with a simple conceptual soundness proof immediately.  I was wrestling with scenarios such as the following:

fun f<T>(...) 
requires exists t1, t2, t3 in T.  t1, t2, t3 are distinct
ensures ...
{ ... }

fun g() {  f<bool>( ... ) }

The precondition of f is clearly inconsistent if T is interpreted to be bool (but not if T is abstract).  On the other hand, when g calls f, the precondition will fail to be justified.  So, everything seems fine.  Still, I was worried if there are some other demons.

File: language/move-prover/bytecode/tests/mono_analysis/test.exp
Comment (wrwg) [reply language/move-prover/bytecode/tests/mono_analysis/test.exp:72]: Yes, type parameter. My hypothesis is right now that I can turn all `#N` types simply into global given sorts which are valid across all functions.

File: language/move-prover/bytecode/src/function_target_pipeline.rs
Comment (wrwg) [reply language/move-prover/bytecode/src/function_target_pipeline.rs:23]: This prepares for FunctionVariant more than just a ground enum.

File: language/move-prover/bytecode/src/mono_analysis.rs
Comment (wrwg) [reply language/move-prover/bytecode/src/mono_analysis.rs]: I renamed it, but this convention you mention I only know for escaping keywords, so choose just a prettier name ;-)

**Comment by wrwg** [comment]
> I could not fully understand the golden output in test.exp (see my question there). But I read through the code in mono_analysis.rs and it made sense to me.
> 
> I thought for a bit about the verification scheme you have in mind where a generic function is verified for an arbitrary uninterpreted type. It seems correct to me intuitively although I could not come up with a simple conceptual soundness proof immediately. I was wrestling with scenarios such as the following:
> 
> fun f(...)
> requires exists t1, t2, t3 in T. t1, t2, t3 are distinct
> ensures ...
> { ... }
> 
> fun g() { f( ... ) }
> 
> The precondition of f is clearly inconsistent if T is interpreted to be bool (but not if T is abstract). On the other hand, when g calls f, the precondition will fail to be justified. So, everything seems fine. Still, I was worried if there are some other demons.

Appreciate the input. We'll iterate on this one. The new mono backend will live side-by-side with the current one for a while.

**Comment by rexhoffman** [comment]
Move is in a broken state, trying to get it back to working again in CI, please refrain from landing until I do.   This is not tested in land.   I may have to revert the last 30 or so commits to figure out what's going on.

**Review by rexhoffman — CHANGES_REQUESTED** [review]
Please refrain from landing more move changes on top of the broken move environment.  Until I get is sorted /land isn't honoring test results.

**Comment by wrwg** [comment]
> Please refrain from landing more move changes on top of the broken move environment. Until I get is sorted /land isn't honoring test results.

Alright let me know when its good again. This PR BTW failed landing just in very regular ways: the merged result with another PR did not compile, and the CI result was from before that other PR landed.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A boogie-backend-exp 0.1.0 (workspace, path 'language/move-prover/boogie-backend-exp')
    * features: [none]
```

**Comment by mengxu-fb** [comment]
/land

(on behalf of @wrwg)

**Pull request closed without merging**
